KAFKA-3430: Allow users to set key in KTable.toStream and in KStream. …

bbejeck opened the pull request:
… With KStream the method selectKey was added to enable getting a key from values before perfoming aggregation-by-key operations on original streams that have null keys.

bbejeck committed:
KAFKA-3420 Allow users to set key in KTable.toStream and in KStream.  With KStream the method selectKey was added to enable getting a key from values before perfoming aggregation-by-key operations on original streams that have null keys.

bbejeck commented:
Commit comment should state "KAFKA-3430"

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKeySelector.java:
```diff
@@ -0,0 +1,31 @@
+
+
+/**
+ * The KStreamKeySelector interface is for pulling a key out of an original stream that does not contain keys
```
Do we need a separate KStreamKeySelector? Or could we just use Mapper? Also it helps in some general cases where users want to select a new key based on both key and value, not just the value.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKeyMapper.java:
```diff
@@ -0,0 +1,31 @@
+ * @param <K1> original key type
+ * @param <K2> mapped key type
+ */
+public interface KTableKeyMapper<K1, K2> {
```
Ditto as above, could we use a KeyValueMapper here?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -244,6 +246,23 @@ public V apply(Change<V> change) {
 
+    @Override
+    public <K1> KStream<K1, V> toStream(KTableKeyMapper<K, K1> kTableKeyMapper) {
+        String toStreamName = topology.newName(TOSTREAM_NAME);
```
As for implementation, we could just call toStream().selectKey(), which will add two processors instead of one, but maybe more illustrative on the topology.

bbejeck committed:
KAFKA-3430 removed interfaces KStreamKeySelector and KTableKeyMapper
and associated processors, updated to use KeyValueMapper.  Updated unit tests.

bbejeck commented:
@guozhangwang  
Changes made to reflect comments:
- removed unnecessary interfaces and processors in favor of using KeyValueMapper and associated processor.
- updated implementation of KTable.toStream(mapper)

bbejeck committed:
KAFKA-3430 changed name of topics in unit-tests.

bbejeck commented:
Not convinced it was the cause of the test failure , but I updated the names of the topics in the unit tests to be more unique.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -46,6 +46,16 @@
 
+
+    /**
+     * Create a key type from the values or a combination of key and value.  Intended to be used before performing
```
nit: "Create a new key from the current key and value."

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -203,6 +203,15 @@
+     * @param mapper  @param mapper  the instance of {@link KeyValueMapper}
+     * @param <K1> the new key type
+     */
+    <K1> KStream<K1, V> toStream(KeyValueMapper<K, V, KeyValue<K1, V>> mapper);
```
Ditto above: "... to select the new key". And use `KeyValueMapper<K, V, K1>`.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -46,6 +46,16 @@
+     * @param mapper  the instance of {@link KeyValueMapper}
+     * @param <K1>   the new key type on the stream
+     */
+    <K1> KStream<K1, V> selectKey(KeyValueMapper<K, V, KeyValue<K1, V>> mapper);
```
Could this be just `KeyValueMapper<K, V, K1>`?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -121,6 +123,13 @@ public KStreamImpl(KStreamBuilder topology, String name, Set<String> sourceNodes
     @Override
+    public <K1> KStream<K1, V> selectKey(KeyValueMapper<K, V, KeyValue<K1, V>> mapper) {
+        String name = topology.newName(KEY_SELECT_NAME);
+        topology.addProcessor(name, new KStreamMap<>(mapper), this.name);
```
With the above change, this would be `new KStreamMap<>(new KeyValueMapper<K, V, KeyValue<K1, V>>) { apply(k, v -> new KeyValue(mapper.apply(k, v), v))}`

bbejeck committed:
KAFKA-3430 changed signature of selectKey and toStream to be KeyValueMapper<K, V, K1> updated implementations and unit tests.  Updated java-docs per suggestions.

bbejeck commented:
@guozhangwang 
   updated per comments.

guozhangwang commented:
Jenkins failure is an known issue, @hachikuji could you take a look?

guozhangwang commented:
LGTM. Merged to trunk, thanks!
